RFC: Range size approximation and key count estimation (#1220)
slatedb/slatedb: rfcs/0018-range-metadata.md

Reviewer: Ah, yes that makes sense. I was actually not referring to those methods because I relied on the rust docs of `SstFile::open()`. So that means only a range_get() is needed. 
Thinking about it again, while it is good to mention what can be reused, for the RFC it does not really matter what requests are issued at creation time of the `SstFile` since it is actually an implementation detail that should be discussed on the implementation PR, IMO.
I would just remove the infos about how many and what requests are needed.

## Goals

- Expose per-SST and per-memtable metadata for a given key range
- Provide a `SizeEstimate` trait with `bytes_before(key)` / `bytes_after(key)` for block-level size estimation within individual SSTs and memtables
- Add an SST stats block to the footer for key/value counts and sizes
- Expose the metadata API on `Db`, `DbReader`, `admin.rs`, and `slatedb-cli`

## Non-Goals

- Exact size calculation or key counting (would require reading and processing a lot more data)
- High-level aggregation functions with configuration options (users derive these from the metadata)

## Design

### API

The following API will be added to `Db`:

```rust
impl Db {
    /// Returns per-SST and per-memtable metadata for the given key range.
    pub async fn metadata<K, T>(
        &self,
        range: T,
    ) -> Result<RangeMetadata, Error>
    where
        K: AsRef<[u8]> + Send,
        T: RangeBounds<K> + Send;
}

pub struct RangeMetadata {
    pub memtables: Vec<MemtableMetadata>,

>     pub ssts: Vec<SstMetadata>,

Reviewer 2: Maybe we should have:

```rs
    pub memtables: Vec<MemtableMetadata>,
    pub l0: Vec<SstMetadata>,
    pub sorted_runs: Vec<SortedRunMetadata>,
    // or maybe HashMap<u32, SortedRunMetadata>?
    // or maybe HashMap<u32, Vec<SstMetadata>>?
```

Grouping SSTs by SR is probably more helpful to users in many cases.

- **`slatedb-cli`**: For command-line operational use cases

### `SizeEstimate` Trait

Both `SstMetadata` and `MemtableMetadata` implement the `SizeEstimate` trait, providing a uniform interface for estimating bytes before/after a key within a single storage unit:

```rust
/// Trait for estimating bytes before/after a key within a single storage unit
/// (SST or memtable). Both methods return overestimates.
pub trait SizeEstimate {
    /// Overestimate of bytes before `key` in this storage unit.
    async fn bytes_before(&self, key: &[u8]) -> Result<u64, Error>;
    /// Overestimate of bytes after `key` in this storage unit.
    async fn bytes_after(&self, key: &[u8]) -> Result<u64, Error>;
}

impl SizeEstimate for SstMetadata { /* reads index block via TableStore */ }
impl SizeEstimate for MemtableMetadata { /* iterates skip map, exact on uncompressed data */ }
```

### Internal State for `SizeEstimate`

**`SstMetadata`** needs to read the SST's index block to implement `bytes_before`/`bytes_after`. For this it holds a private `Arc<TableStore>` — the existing `pub(crate)` struct that reads index blocks from object storage, using the block cache when available.

> Currently `TableStore::read_index` takes a full `&SsTableHandle`, but it only uses `handle.id` (to resolve the object store path and cache key) and `handle.info` (for `index_offset`/`index_len`/`compression_codec`). It does not use the handle's `visible_range` or `effective_range`. A small refactor to accept `(id: &SsTableId, info: &SsTableInfo)` instead would make the dependency cleaner. Since `SstMetadata` already stores `id`, `first_key`, and `compression_codec` as public fields, the only additional private state needed is `index_offset` and `index_len` (stored internally via `SsTableInfo`).

Author: This is Claude who came up with this approach. I don't know if there should be some minor refactor to better obtain the info needed here via `TableStore::read_index` without also storing the full `SsTableInfo` on `SstMetadata`.

    /// `get_range()` call). Errors if the file was GC'd.
    pub async fn open(&self, id: Ulid) -> Result<SstFile, crate::Error>;
}
```

`SstReader` wraps the read-only functionality currently internal to `TableStore`. SST paths are resolved as `{root}/compacted/{ulid}.sst`. Internally, `open()` needs to decode the SST footer and index block, which requires handling the SST codec and compression. The implementation will need to take care of constructing the appropriate decoding machinery (currently `SsTableFormat`) from the provided configuration. If `object_store_cache_options` is provided, the passed-in `object_store` is wrapped in a `CachedObjectStore` (mirroring the behavior of `DbBuilder`). This allows the `SstReader` to share the on-disk object store cache that `DbBuilder` sets up internally. Note that this cache path can be shared between processes.

```rust
pub struct SstFile {
    id: Ulid,
    handle: SsTableHandle,
    tablestore: Arc<TableStore>,
}

impl SstFile {
    /// Returns the SST's ULID identifier.
    pub fn id(&self) -> Ulid;

    /// Returns the SST file metadata (requires a `head()` call to the object store).

>     pub async fn metadata(&self) -> Result<SstFileMetadata, crate::Error>;

Reviewer: In `WalFile` the result is `Option<WalFileMetadata>` since a WAL could have been removed. 
Can a SST file also not exist anymore due to compaction?

4. **User reads** → `db.metrics().lookup("db.memtable_num_puts")` returns the current gauge value, representing the aggregate across the active memtable plus all immutable memtables not yet flushed.

Note that these stats are global, not per-range. You cannot ask "how many puts are in memtables for key range X..Y". This is an accepted limitation since memtables are bounded in size (by `max_unflushed_bytes`) and the SST-level stats dominate for most estimation use cases.

### Size and Cardinality Estimation

This section describes how to use the above APIs for common estimation tasks, at three levels of accuracy.

#### Coarse estimate — SST-level, no I/O beyond `manifest()`

Call `db.manifest()`. For each sorted run, use `tables_covering_range(range)` to find the intersecting SSTs. For L0, iterate all handles (L0 SSTs can overlap arbitrarily). Sum `estimate_size()` across all covering SSTs. This gives an upper-bound stored-bytes estimate. No `SstReader` needed. Overestimates because boundary SSTs are counted in full even if the range only touches a small portion.

> For cardinality: open each covering SST with `SstReader` and call `SstFile::info()` to get `SsTableInfo`. Sum `num_puts + num_merges - num_deletes` across all levels. This overestimates because the same key may appear in multiple levels (L0 overlaps, and compaction hasn't merged them yet). The overestimation factor is roughly proportional to the number of levels containing the key.

Reviewer 2: > I am also not understanding why you need to get SsTableInfo through SstFile::info(). Calling tables_covering_range(range) returns a queue of SsTableHandle which should contain the SsTableInfos.

Yes, for @FiV0's use case, no `stats()` call is nseeded--the stats are available via a single `db.manifest()` call. The `SstFile::stats()` call would just bee for users that don't have access to a `ManifestCore` (e.g. if they just listed files in object storage outside of SlateDB and wanted to inspect them). The `SstFile::index()` is needed for @FiV0 iff sub-SST sized approximations are required.

> For cardinality: open each covering SST with `SstReader` and call `SstFile::info()` to get `SsTableInfo`. Sum `num_puts + num_merges - num_deletes` across all levels. This overestimates because the same key may appear in multiple levels (L0 overlaps, and compaction hasn't merged them yet). The overestimation factor is roughly proportional to the number of levels containing the key.

Reviewer 2: Regarding the head request, I don't think that's issued AFAICT. If we use TableStore to open the SST, I think it just fetches the metadata block (`SsTableInfo`) and returns the handle. So that would be one range get().

> For cardinality: open each covering SST with `SstReader` and call `SstFile::info()` to get `SsTableInfo`. Sum `num_puts + num_merges - num_deletes` across all levels. This overestimates because the same key may appear in multiple levels (L0 overlaps, and compaction hasn't merged them yet). The overestimation factor is roughly proportional to the number of levels containing the key.

Reviewer 2: I don't think we can encode it in the SST itself if we want to support things like trivial moves:

https://github.com/slatedb/slatedb/issues/1110

Since the SST could be moved between SRs.

- Provide `SstReader` and `SstFile` for opening individual SSTs and reading their stats and index data
- Expose memtable stats as metrics via the existing `StatRegistry` / `Db::metrics()`

## Non-Goals

- Exact size calculation or key counting (would require reading and processing a lot more data)
- High-level aggregation functions with configuration options (users derive these from the metadata)
- Providing a single `Db::metadata(range)` API that pre-computes which SSTs overlap a range (users do this themselves using the manifest)
- Exposing `SstFile::rows()` for reading row data (out of scope for size estimation)

## Design

### API

#### `Db::manifest()`

```rust
impl Db {
    ...
    pub fn manifest(&self) -> ManifestCore;
    ...
}
```

Returns a clone of the current in-memory `ManifestCore`. No I/O. `ManifestCore` is already public (re-exported via `slatedb::manifest`) .

#### `SstReader` and `SstFile`

```rust
pub struct SstReader {
    object_store: Arc<dyn ObjectStore>,
    root_path: Path,
    cache: Option<Arc<dyn DbCache>>,
}

impl SstReader {
    pub fn new(
        root_path: impl Into<Path>,
        object_store: Arc<dyn ObjectStore>,

>         cache: Option<Arc<dyn DbCache>>,

Reviewer 2: We might also want to take an object store cache path here. See `object_store_cache_options` in `DbBuilder`. IIRC, this path can be shared between processes (cc @flaneur2020).

Author: I can add it if you want. Will users not very likely just pass in the object store they configure via `DbBuilder` and then it's wrapped or not depending on that configuration? I mean it comes back to how this object should be initialized. Should it just use everything as it's used in DbBuilder or should one be able to configure things via a more specialized constructor such as this. Personally, I'd just like to get a `SstReader` from a `Db`/ `DbConfig`.

Reviewer 2: > Will users not very likely just pass in the object store they configure via DbBuilder and then it's wrapped or not depending on that configuration.

They pass the object store into DbBuilder, and then DbBuilder wraps the object store internally in a cached object store. The user doesn't have access to the cached object store, unfortunately. So we'd have to emulate that here: take an `Option<ObjectStoreCacheOptions>` and if it's set, wrap the passed-in object_store in a `CachedObjectStore`.

> Personally, I'd just like to get a SstReader from a Db/ DbConfig.

@agavra just raised a very similar comment about `db.wal_reader()`. @agavra Maybe we should just go for it for both? 🤔

Reviewer 2: I'm leaning toward just keeping the these 5 stats in the `SsTableInfo` block instead of a separate stats block. As you said, since we're doing row-based, not columnar (analytics) stuff, it should be fewer stats. I looked at RocksDB's SST stats and they more or less map to what we're talking about.

    ) -> Self;

    /// Reads the SST footer from object storage (one `head()` + one
    /// `get_range()` call). Errors if the file was GC'd.
    pub async fn open(&self, id: Ulid) -> Result<SstFile, SlateDBError>;
}
```

> `SstReader` wraps the read-only functionality currently internal to `TableStore`. SST paths are resolved as `{root}/compacted/{ulid}.sst`. Internally, `open()` needs to decode the SST footer and index block, which requires handling the SST codec and compression. The implementation will need to take care of constructing the appropriate decoding machinery (currently `SsTableFormat`) from the provided configuration.

Reviewer 2: Thinking about this more, I think it's fine if WalFile and SstFile behave differently. If anything, I think the open() API is more intuitive/less surprising. I might actually align the WalFile to behave more like it. In any case, I retract my earlier suggestion.

```rust
pub struct SstFile {
    id: Ulid,
    info: SsTableInfo,
    stats: SstStats,
    reader: SstReader,
}

impl SstFile {
    pub fn id(&self) -> Ulid;

> 

Reviewer 2: Might want to expose `metadata()` here as well (similar to WalFile).

}
```

`index()` calls `SsTableFormat::read_index()`, which reads `info.index_offset..info.index_offset + info.index_len`, decompresses, and returns an `SsTableIndexOwned`. The method materializes `Vec<(u64, Bytes)>` from the FlatBuffer `BlockMeta` entries (each has `offset()` and `first_key()`). Caching uses `DbCache::get_index` / `insert` keyed by `(sst_id, index_offset)`, matching the existing pattern in `TableStore::read_index()`.

#### `SstStats`

Per-SST statistics from the stats block in the footer (Phase 1). For pre-Phase-1 SSTs, all fields are 0.

```rust
pub struct SstStats {
    pub num_puts: u64,
    pub num_deletes: u64,
    pub num_merges: u64,
    pub size_raw_key_bytes: u64,
    pub size_raw_value_bytes: u64,
}
```

#### Visibility changes to existing types

- **`SortedRun::tables_covering_range()`**: Change from `pub(crate)` to `pub`. Lets users find which SSTs in a sorted run overlap a given key range.
- **`SsTableHandle::estimate_size()`**: Change from `pub(crate)` to `pub`. Enables quick no-I/O size estimates.
- **`SsTableHandle::visible_range`**: Consider changing from `pub(crate)` to `pub` so users can inspect the projected key range.

> #### Memtable stats via `StatRegistry`

Reviewer 2: This section seems solid to me. 👍

#### L0 handling

L0 SSTs are not sorted relative to each other and can all overlap with the query range. Every L0 SST must be checked. For each, the same refinement as above applies: if the SST's `first_key` and key range suggest partial overlap, use `index()` to refine. Since L0 SSTs tend to be smaller (pre-compaction), the coarse estimate is often sufficient.

#### Staleness and GC

The manifest snapshot from `manifest()` may diverge from the actual DB state by the time `SstReader::open()` is called. GC may have deleted SSTs. Treat `open()` errors on missing SSTs as best-effort — skip them. For stronger guarantees, create a checkpoint before reading.

#### Memtable contribution

Add memtable stats from `db.metrics()` (`memtable_num_puts`, `memtable_raw_key_bytes`, etc.) to the totals. These are aggregates across all unflushed memtables and are not range-specific.

### Implementation Phases

#### Phase 1 - SST Stats Block

Add a stats block to the SST footer with the following fields (all `u64`):
- `num_puts`
- `num_deletes`
- `num_merges`
- `raw_key_size`
- `raw_val_size`

> These are loaded into `SsTableInfo` at SST open time. This requires care as the storage format changes, i.e. the `.fbs` gets updated. For backwards compatibility, the stats block goes at the end of the footer and is optional. Old SSTs without a stats block get zeros for these fields.

Reviewer 2: I'm leaning toward just keeping the these 5 stats in the SsTableInfo block instead of a separate stats block. As you said, since we're doing row-based, not columnar (analytics) stuff, it should be fewer stats. I looked at RocksDB's SST stats and they more or less map to what we're talking about.

- The `error_margin` approach sometimes triggered I/O and sometimes didn't, making performance unpredictable

The immediately preceding revision proposed a `Db::metadata(range)` API returning `RangeMetadata` with `Vec<SstMetadata>` and `Vec<MemtableMetadata>`, plus a `SizeEstimate` trait with `bytes_before(key)` / `bytes_after(key)` methods. This was replaced because:
- It coupled range-walking logic to the DB, when users could do this themselves using the manifest
- The `SizeEstimate` trait with private `Arc<TableStore>` state leaked internal abstractions
- Memtable metadata required holding `Arc<KVTable>` references, complicating lifecycle management
- The new approach is more modular: `SstReader`/`SstFile` are independent of `Db` and can be used with any `ObjectStore` and optional `DbCache`

Another alternative not explored is sample-based estimation: sample N random blocks and extrapolate key counts or compression ratios. This could complement the current approach but adds complexity.

## Open Questions

- Should `Db` expose a pre-configured `SstReader` via `db.sst_reader()` instead of requiring manual construction?

> - Should `SstFile::index()` return raw `SsTableIndexOwned` instead of materializing `Vec<(u64, Bytes)>` to avoid the allocation?

Author: Let's keep it this way.

## Goals

- Add an SST stats block to the footer for key/value counts and sizes
- Expose `Db::manifest()` so users can discover SSTs covering a key range
- Provide `SstReader` and `SstFile` for opening individual SSTs and reading their stats and index data
- Expose memtable stats as metrics via the existing `StatRegistry` / `Db::metrics()`

## Non-Goals

- Exact size calculation or key counting (would require reading and processing a lot more data)
- High-level aggregation functions with configuration options (users derive these from the metadata)
- Providing a single `Db::metadata(range)` API that pre-computes which SSTs overlap a range (users do this themselves using the manifest)
- Exposing `SstFile::rows()` for reading row data (out of scope for size estimation)

## Design

### API

#### `Db::manifest()`

```rust
impl Db {
    ...
    pub fn manifest(&self) -> &ManifestCore;
    ...
}
```

> Returns a clone of the current in-memory `ManifestCore`. No I/O. `ManifestCore` is already public (re-exported via `slatedb::manifest`) .

Reviewer 2: ```suggestion
Returns a reference to the current in-memory `ManifestCore`. No I/O. `ManifestCore` is already public (re-exported via `slatedb::manifest`) .
```

#### `SstReader` and `SstFile`

```rust
pub struct SstReader {
    object_store: Arc<dyn ObjectStore>,
    root_path: Path,
    cache: Option<Arc<dyn DbCache>>,
}

impl SstReader {
    pub fn new(
        root_path: impl Into<Path>,
        object_store: Arc<dyn ObjectStore>,
        cache: Option<Arc<dyn DbCache>>,
    ) -> Self;

    /// Reads the SST footer from object storage (one `head()` + one
    /// `get_range()` call). Errors if the file was GC'd.
    pub async fn open(&self, id: Ulid) -> Result<SstFile, SlateDBError>;
}
```

`SstReader` wraps the read-only functionality currently internal to `TableStore`. SST paths are resolved as `{root}/compacted/{ulid}.sst`. Internally, `open()` needs to decode the SST footer and index block, which requires handling the SST codec and compression. The implementation will need to take care of constructing the appropriate decoding machinery (currently `SsTableFormat`) from the provided configuration.

```rust

> pub struct SstFile {

Reviewer 2: If we keep the stats in SsTableInfo, I think we should do:

```rs
pub struct SstFile {
    id: Ulid,
    handle: SsTableHandle,
    tablestore: Arc<TableStore>,
}

impl SstFile {
    /// Returns the SST's ULID identifier.
    pub fn id(&self) -> Ulid;

    /// Returns the SST file metadata (requires a `head()` call to the object store).
    pub async fn metadata(&self) -> Result<SstFileMetadata, crate::Error>;

    /// Returns SsTableInfo from the metadata block.
    pub fn info(&self) -> Result<SsTableInfo, crate::Error>;

    /// Returns `(block_offset, first_key)` pairs from the SST index block.
    pub async fn index(&self) -> Result<Vec<(u64, Bytes)>, crate::Error>;
}
```

This has a few benefits:

1. We don't need a new SstStats struct. SsTableInfo contains it.
2. Caching handle means we can use `tablestore.read_index`, which requires an SsTableHandle.

The downside is that this approach requires a read to get SsTableHandle even if all we want to do is call metadata(), which doesn't require it. I think this is a fine tradeoff.
